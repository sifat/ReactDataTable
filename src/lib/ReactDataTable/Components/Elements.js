import { React, useContext, useEffect } from 'react';
import { DataContext } from '../Contexts';
import loading_icon from '../Loading_icon.gif';

export function Loading({ numColumn }) {
    return (
        <tr className="tr-loading">
            <td colSpan={numColumn}>
                <div style={{'textAlign': 'center'}} >
                    <img src={loading_icon} alt="Loading..." style={{'width': '100px', 'height': '70px'}} />
                </div>
            </td>
        </tr>
    )
}

/**
 * 
 * @param {type} border, grow, auto
 * @param {color} 
 *  text color utilities
 * text-primary,text-secondary,text-success,text-danger,text-warning,text-info, text-light, text-dark  
 * @param margin margin utilities 
 */
export function BootStrapSpinners({ className, color, margin, size }) {
    
    if (size && size == 'small' && (className.indexOf('spinner-border') >= 0  || className.indexOf('ms-auto') >= 0 )) {
        className += ' spinner-border-sm ';
    }
    if (size && size == 'small' && className.indexOf('spinner-grow') >= 0) {
        className += ' spinner-grow-sm ';
    }
    if (color) {
        className += ` ${color} `;
    }
    if (margin) {
        className += ` ${margin} `;
    }
    
    return (
        <div className={className} role="status">
            <span className="visually-hidden">Loading...</span>
        </div>
    );
}

/**
 * 
 * @param {color} 
 *  text color utilities  
 * @param {margin} margin utilities
 * @param {size} small
 */
export function BootStrapSpinnerBorder(props) {
    return <BootStrapSpinners className="spinner-border" {...props} />
}

/**
 * 
 * @param {color} 
 *  text color utilities  
 * @param {margin} margin utilities
 * @parma {size} small
 */
export function BootStrapSpinnerGrow(props) {
    return <BootStrapSpinners className="spinner-grow" {...props} />
}

/**
 * 
 * @param {textAlignClassName} text alignment utilities
 * @param {className} class of spinner 
 * @returns 
 */
export function BootStrapSpinnerTextAlign(props) {
    return (
        <div className={props.textAlignClassName}>
            <BootStrapSpinners {...props} />
        </div>
    );
}

/**
 * 
 * @param {floatAlignment} float utilities
 * @param {className} class of spinner 
 * @returns 
 */
export function BootStrapSpinnerFloat(props) {
    // const className = `${props.floatAlignment} ${props.className}`;
    return (
        <div className="clearfix">
            <BootStrapSpinners {...props} />
        </div>
    );
}

/**
 * 
 * @param {flex} flexbox utilities
 * @param {className} class of spinner 
 * @returns 
 */
export function BootStrapSpinnerFlex(props) {
    return (
        <div className={props.flex}>
            <BootStrapSpinners {...props} />
        </div>
    );
}





export function ReactTableCheckBox({ name, value, checked, className }) {
    const data = useContext(DataContext);
    useEffect(() => {
        if (checked) data.setCheckbocStatus(checked)    
    }, []);
    return (
        <input type="checkbox" 
        name={name} 
        className={className} 
        value={value} 
        checked={data.checkboxStatus}
        onChange={() => {
            data.setCheckbocStatus((status) => !status)    
        }} 
        />
    );
}
