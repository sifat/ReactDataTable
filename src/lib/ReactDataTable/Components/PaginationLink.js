import { useContext } from "react";
import { DataContext, OptionContext } from "../Contexts";
import { pagintateData } from "../utils";

export default function PaginationLink({num}) {
    const option = useContext(OptionContext);
    const data = useContext(DataContext);

    const isActive = num == data.appState.currentPage;

    return (
        <li className={`${option.pagination.pageItemClassName} ${isActive ? 'active' : ''}`} onClick={() => {
            if (isActive) return true;

            data.setAppState((prev) => {
                return {
                    ...prev,
                    currentPage: num
                }
                
            });

            if (option.remote) {
                return true;
            }
            data.setIsLoading(true);
            data.setData(pagintateData([...data.filteredData.current], option.pagination.perPage, num));
            data.setIsLoading(false);
        }}>
            <a className={option.pagination.anchorClassName} >{num}</a>
        </li>
    );
}