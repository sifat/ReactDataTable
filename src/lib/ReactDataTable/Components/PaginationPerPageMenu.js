import { useState, useContext } from "react";
import { DataContext, OptionContext } from "../Contexts";
import { pagintateData } from "../utils";

export default function PaginationPerPageMenu() {
    const options = [];
    const data = useContext(DataContext);
    const option = useContext(OptionContext);
    const [value, setValue] = useState(data.appState.perPage);
    const style = JSON.parse(option.pagination.lengthMenu.style);
    style['float'] = option.pagination.lengthMenu.position == 'right' ? 'right' : 'left';

    for (const op of option.pagination.lengthMenu.menu) {
        // console.log(option);
        options.push(<option value={op} key={op}>{op}</option>);
    }



    return (
        <div className={option.pagination.lengthMenu.wrapperClassName} style={JSON.parse(option.pagination.lengthMenu.style)}>
            <label className={option.pagination.lengthMenu.labelClassName}>Show</label>
            <select name='perPage' value={value} className={option.pagination.lengthMenu.className} onChange={(e) => {
                setValue(e.target.value);
                data.setAppState((prev) => {
                    return {
                        ...prev,
                        perPage: e.target.value,
                        currentPage: 1
                    }
                });
                if (option.remote) {
                    return true;
                }
                const totalPage = Math.ceil(data.filteredData.current.length / e.target.value);
                data.setTotalPage(totalPage);
                data.setIsLoading(true);
                data.setData(pagintateData([...data.filteredData.current], e.target.value, 1));
                data.setIsLoading(false);
            }}>
                {options}
            </select>
            <label className={option.pagination.lengthMenu.labelClassName}>entries</label>

        </div>
    );
}
