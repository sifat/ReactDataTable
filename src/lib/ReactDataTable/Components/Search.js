import { useContext } from "react";
import { DataContext, NoResultFoundContext, OptionContext } from "../Contexts";
import { pagintateData, initalProps } from "../utils";

export default function Search() {
    const option = useContext(OptionContext);
    const data = useContext(DataContext);
    const noResult = useContext(NoResultFoundContext);
    
    let props = initalProps(option, 'search');
    return (
        <div className={option.search.wrapperClassName}>
            <input type={"text"} {...props} name={option.search.name} onKeyUp={(e) => {
                
                if (option.remote) {
                    // getRemoteData(option, data);
                    data.setAppState((prev) => {
                        return {
                            ...prev,
                            keyword: e.target.value,
                            currentPage: 1
                        }
                    });
                    return true;
                }

                
                if (noResult.noResultFound === true) {
                    noResult.setNoResultFound(false);
                } 

                data.setIsLoading(true);
                let newData = [];
                if (e.target.value.length == 0) {
                    newData = data.original;

                } else {
                    let keyword = option.search.casesensitive ? e.target.value : String(e.target.value).toLocaleLowerCase();
                    for (const i in data.original) {
                        for (const j in data.original[i]) {
                            let fieldValue = String(data.original[i][j]);
                            if (!option.search.casesensitive) {
                                fieldValue = fieldValue.toLocaleLowerCase();
                            }
                            if (fieldValue.indexOf(keyword) >= 0) {
                                newData.push(data.original[i]);
                                break;
                            }
                        }
                    }
                }


                data.filteredData.current = newData;
                const totalPage = Math.ceil(newData.length / option.pagination.perPage);
                data.setAppState((prev) => {
                    return {
                        ...prev,
                        keyword: e.target.value,
                        currentPage: 1
                    }
                });
                data.setTotalPage(totalPage);
                if (option.pagination.enablePagination) {
                    newData = pagintateData(newData, option.pagination.perPage, 1);
                }
                data.setData(newData);
                data.setIsLoading(false);
                if (newData.length == 0) {
                    noResult.setNoResultFound(true);
                }
            }} />
        </div>
    );
}