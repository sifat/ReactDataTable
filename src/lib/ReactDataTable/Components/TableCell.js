export default function TableCell({ data, column }) {
    let cell = data[column.dataIndex];
    // console.log();
    if (column.hasOwnProperty('render') && typeof (column.render) == 'function') {
        cell = column.render(data);
    }

    return (
        <td>{cell}</td>
    )
}