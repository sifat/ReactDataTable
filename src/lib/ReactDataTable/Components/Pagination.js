import { useContext } from "react";
import { DataContext, OptionContext } from "../Contexts";
import PaginationLink from "./PaginationLink";

export default function Pagination() {
    const option = useContext(OptionContext);
    const data = useContext(DataContext);

    if (!option.pagination.enablePagination) return null;

    // console.log(data.totalPage);
    const paginationLinks = [];
    for (let i = 0; i <= data.totalPage + 1; i++) {
        paginationLinks.push(<PaginationLink num={i} key={i} />);
    }
    return (
        <nav className={option.pagination.wrapperClassName}>
            <ul className={option.pagination.className}>
                {paginationLinks}
            </ul>
        </nav>
    );
}