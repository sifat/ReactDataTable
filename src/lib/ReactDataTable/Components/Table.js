import { useContext } from "react";
import { DataContext, OptionContext } from "../Contexts";
import Search from "./Search";
import { initalProps } from "../utils";

export default function Table({ children }) {
    const option = useContext(OptionContext);
    const props = initalProps(option, 'table');
    
    const search = option.search.enableSearch ? <Search /> : null;
    let heading = null;
    if (option.heading.enableHeading) {
        const title = <div className={option.heading.titleWrapperClassName}>
            {option.heading.title}
        </div>
        const rightContent = option.search.position == 'right' ? search : title;
        const leftContent = option.search.position == 'left' ? search : title;
        heading = <div className={option.heading.className}>
            {leftContent}
            {rightContent}
        </div>
    }
    return (
        <>
            {heading}
            <table {...props}>
                {children}
            </table>
        </>
    );
}