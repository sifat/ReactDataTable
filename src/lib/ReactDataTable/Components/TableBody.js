import { useContext } from "react";
import { ColumnContext, DataContext, NoResultFoundContext, OptionContext } from "../Contexts";
import TableCell from "./TableCell";

export default function TableBody({loader}) {
    const data = useContext(DataContext);
    let columns = useContext(ColumnContext);
    const noResult = useContext(NoResultFoundContext);
    const option = useContext(OptionContext);
    
    if (data.isLoading) {
        return (
            <tbody>
                {loader}
            </tbody>
        )
    }

    if (noResult.noResultFound) {
        return (
            <tbody>
                <tr>
                    <td colSpan={columns.length}>
                        <div className="no-result-found">
                            {option.search.notFoundText}
                        </div>
                    </td>
                </tr>
            </tbody>
        );
    }
    
    const tr = data.current.map((row, index) => {
        let contents = '';
        const td = columns.map((column) => {
            contents += row[column.dataIndex];
            return (
                <TableCell data={row} column={column} key={column.dataIndex} />
            )
        });
        return (
            <tr key={btoa(contents)}>
                {td}
            </tr>
        );
    });
    return (
        <tbody>
            {tr}
        </tbody>
    );
}