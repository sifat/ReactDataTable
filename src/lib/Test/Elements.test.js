import renderer from 'react-test-renderer';
import { BootStrapSpinnerBorder, BootStrapSpinners, Loading } from '../ReactDataTable/Components/Elements';

it('BootStrapSpinners test', () => {
    const spinner = renderer.create(
        <BootStrapSpinners className="spinner-border" color="text-primary" size="small" />
    );
    let tree = spinner.toJSON();
    expect(tree).toMatchSnapshot();
});

it('BootStrapSpinnerBorder test', () => {
    const spinner = renderer.create(
        <BootStrapSpinnerBorder color="text-primary" margin="m-5" />
    );
    let tree = spinner.toJSON();
    expect(tree).toMatchSnapshot();
});

it('Default LoaderTest', () => {
    const loader = renderer.create(
        <Loading numColumn={2} />
    );
    let tree = loader.toJSON();
    expect(tree).toMatchSnapshot();
});