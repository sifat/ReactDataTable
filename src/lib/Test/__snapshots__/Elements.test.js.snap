// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BootStrapSpinnerBorder test 1`] = `
<div
  className="spinner-border text-primary  m-5 "
  role="status"
>
  <span
    className="visually-hidden"
  >
    Loading...
  </span>
</div>
`;

exports[`BootStrapSpinners test 1`] = `
<div
  className="spinner-border spinner-border-sm  text-primary "
  role="status"
>
  <span
    className="visually-hidden"
  >
    Loading...
  </span>
</div>
`;
